crates/wasi-common/yanix/src/dir.rs

```diff
@@ -0,0 +1,171 @@
+use crate::{
+    sys::dir::{iter_impl, EntryImpl},
+    Errno, Result,
+};
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
+use std::{ffi, ops::Deref, ptr};
+
+pub use crate::sys::EntryExt;
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct Dir(pub ptr::NonNull<libc::DIR>);
+
+impl Dir {
+    /// Converts from a descriptor-based object, closing the descriptor on success or failure.
+    #[inline]
+    pub fn from<F: IntoRawFd>(fd: F) -> Result<Self> {
+        unsafe { Self::from_fd(fd.into_raw_fd()) }
```

Comment:
As a minor stylistic nit, we could split this into two lines, holding the raw fd in a local variable, so that only the `from_fd` is inside the unsafe block, which helps illustrate what this code is doing that's unsafe.